src/flow/src/utils.rs

```diff
@@ -0,0 +1,502 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{BTreeMap, BTreeSet};
+use std::sync::Arc;
+
+use itertools::Itertools;
+use serde::{Deserialize, Serialize};
+use smallvec::{smallvec, SmallVec};
+use tokio::sync::{Mutex, RwLock};
+
+use crate::expr::error::InternalSnafu;
+use crate::expr::{EvalError, ScalarExpr};
+use crate::repr::{value_to_internal_ts, Diff, DiffRow, Duration, KeyValDiffRow, Row, Timestamp};
+
+/// Determine when should a key expire according to it's event timestamp in key
+///
+/// TODO(discord9): find a better way to handle key expiration, like write to disk or something instead of throw away
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd, Deserialize, Serialize)]
+pub struct KeyExpiryManager {
+    /// a map from event timestamp to key, used for expire keys
+    event_ts_to_key: BTreeMap<Timestamp, BTreeSet<Row>>,
+    /// duration after which a key is considered expired, and will be removed from state
+    key_expiration_duration: Option<Duration>,
+    /// using this to get timestamp from key row
+    event_timestamp_from_row: Option<ScalarExpr>,
+}
+
+impl KeyExpiryManager {
+    /// extract event timestamp from key row
+    pub fn extract_event_ts(&self, row: &Row) -> Result<Option<Timestamp>, EvalError> {
+        let ts = self
+            .event_timestamp_from_row
+            .as_ref()
+            .map(|e| e.eval(&row.inner))
+            .transpose()?
+            .map(value_to_internal_ts)
+            .transpose()?;
+        Ok(ts)
+    }
+
+    pub fn get_expire_time(&self, now: Timestamp) -> Option<Timestamp> {
+        self.key_expiration_duration.map(|d| now - d)
+    }
+
+    /// update the event timestamp to key mapping
+    ///
+    /// if given key is expired by now, return false
+    pub fn update_event_ts(&mut self, now: Timestamp, row: &Row) -> Result<bool, EvalError> {
+        let ts = if let Some(event_ts) = self.extract_event_ts(row)? {
+            if self
+                .get_expire_time(now)
+                .map(|e| e > event_ts)
+                .unwrap_or(false)
+            {
+                // expired
+                return Ok(false);
+            }
+            event_ts
+        } else {
+            return Ok(true);
+        };
+
+        self.event_ts_to_key
+            .entry(ts)
+            .or_default()
+            .insert(row.clone());
+        Ok(true)
+    }
+}
+
+/// A shared state of key-value pair for various state
+/// in dataflow execution
+///
+/// i.e: Mfp operator with temporal filter need to store it's future output so that it can add now, and delete later.
+/// To get all needed updates in a time span, use `get_updates_in_range`
```

Comment:
Sugg.: enclose those tokens in `[]` and there will be a link on generated docs. E.g.:
```suggestion
/// To get all needed updates in a time span, use [`get_updates_in_range`]
```